_elseNested: Entity with more than one subfield results in multiple entities

Reported by @hagbeck:

> I've observed another effect.

```xml
<marc:datafield tag="856" ind1="4" ind2="0">
   <marc:subfield code="u">http://www.video2brain.com/</mx:subfield>
   <marc:subfield code="x">Agentur</mx:subfield>
</marc:datafield>
```

results in multiple entities like

```xml
<marc:datafield tag="856" ind1="4" ind2="0">
        <marc:subfield code="u">http://www.video2brain.com/</marc:subfield>
</marc:datafield>
<marc:datafield tag="856" ind1="4" ind2="0">
        <marc:subfield code="x">Agentur</marc:subfield>
</marc:datafield>
```

using the ['_elseNested' morph](https://github.com/metafacture/metafacture-core/issues/336#issue-730486530).

Originates from #107.

Fortunately, we have a test for this. Unfortunately, the test is insufficient. All (?) tests should `verifyNoMoreInteractions`:

```diff
diff --git metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
index a564bed8..29fc828c 100644
--- metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -19,6 +19,7 @@
 import static org.mockito.Mockito.inOrder;
 import static org.mockito.Mockito.times;
 import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.verifyNoMoreInteractions;
 
 import org.junit.Rule;
 import org.junit.Test;
@@ -156,6 +157,8 @@ public void shouldHandleUnmatchedLiteralsAndEntitiesInElseNestedSource() {
         ordered.verify(receiver).literal("Hawaii", "Aloha");
         ordered.verify(receiver).endEntity();
         ordered.verify(receiver).endRecord();
+        ordered.verifyNoMoreInteractions();
+        verifyNoMoreInteractions(receiver);
     }
 
     @Test
```

```none
> Task :metamorph:test

org.metafacture.metamorph.TestMetamorphBasics > shouldHandleUnmatchedLiteralsAndEntitiesInElseNestedSource FAILED
    org.mockito.exceptions.verification.NoInteractionsWanted at TestMetamorphBasics.java:161

168 tests completed, 1 failed, 1 skipped

[...]

BUILD FAILED in 5s
126 actionable tasks: 2 executed, 124 up-to-date
```

```diff
diff --git metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
index 29fc828c..1cf6700d 100644
--- metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -151,6 +151,8 @@ public void shouldHandleUnmatchedLiteralsAndEntitiesInElseNestedSource() {
         ordered.verify(receiver).literal("Shikotan", "Aekap");
         ordered.verify(receiver).startEntity("Germany");
         ordered.verify(receiver).literal("Langeoog", "Moin");
+        ordered.verify(receiver).endEntity();
+        ordered.verify(receiver).startEntity("Germany");
         ordered.verify(receiver).literal("Baltrum", "Moin Moin");
         ordered.verify(receiver).endEntity();
         ordered.verify(receiver).startEntity("USA");
```

```none
BUILD SUCCESSFUL in 5s
131 actionable tasks: 2 executed, 129 up-to-date
```

--- a/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
+++ b/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
@@ -103,7 +103,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
     private boolean elseNestedEntityStarted;
     private String currentLiteralName;
     private static final Logger LOG = LoggerFactory.getLogger(Metamorph.class);
-
+    final private Pattern literalPatternOfEntityMarker = Pattern.compile(flattener.getEntityMarker(), Pattern.LITERAL);
     protected Metamorph() {
         // package private
         init();
@@ -280,7 +280,6 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         ++entityCount;
         currentEntityCount = entityCount;
         entityCountStack.push(Integer.valueOf(entityCount));
-
         flattener.startEntity(name);
     }
 
@@ -318,11 +317,10 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     private void dispatch(final String path, final String value, final List<NamedValueReceiver> fallbackReceiver, final boolean endEntity) {
         final List<NamedValueReceiver> matchingData = dataRegistry.get(path);
-
         if (matchingData != null && !matchingData.isEmpty()) {
             send(path, value, matchingData);
         }
-        else if (fallbackReceiver != null) {
+        else if (fallbackReceiver != null && (literalPatternOfEntityMarker.split(path).length >0 && dataRegistry.get(literalPatternOfEntityMarker.split(path)[0]).isEmpty())) {
             if (endEntity) {
                 if (elseNestedEntityStarted) {
                     outputStreamReceiver.endEntity();
@@ -337,7 +335,6 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
                         outputStreamReceiver.startEntity(entityName);
                         elseNestedEntityStarted = true;
                     }
-
                     send(currentLiteralName, value, fallbackReceiver);
                 }
                 else {

--- a/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
+++ b/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
@@ -122,7 +122,6 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     public Metamorph(final String morphDef, final Map<String, String> vars,
             final InterceptorFactory interceptorFactory) {
-
         this(getInputSource(morphDef), vars, interceptorFactory);
     }
 
@@ -140,7 +139,6 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     public Metamorph(final Reader morphDef, final Map<String, String> vars,
             final InterceptorFactory interceptorFactory) {
-
         this(new InputSource(morphDef), vars, interceptorFactory);
     }
 
@@ -158,7 +156,6 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     public Metamorph(final InputStream morphDef, final Map<String, String> vars,
             final InterceptorFactory interceptorFactory) {
-
         this(new InputSource(morphDef), vars, interceptorFactory);
     }
 
@@ -224,14 +221,16 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     protected void registerNamedValueReceiver(final String source, final NamedValueReceiver data) {
         if (ELSE_NESTED_KEYWORD.equals(source)) {
-            this.elseNested = true;
+            elseNested = true;
         }
+
         if (ELSE_KEYWORD.equals(source) || ELSE_FLATTENED_KEYWORD.equals(source) || elseNested) {
-            if (elseSources.isEmpty())
+            if (elseSources.isEmpty()) {
                 elseSources.add(data);
-            else
-                LOG.warn(
-                        "Only one of '_else', '_elseFlattened' and '_elseNested' is allowed. Ignoring the superflous ones.");
+            }
+            else {
+                LOG.warn("Only one of '_else', '_elseFlattened' and '_elseNested' is allowed. Ignoring the superflous ones.");
+            }
         } else {
             dataRegistry.register(source, data);
         }
@@ -258,7 +257,6 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     @Override
     public void endRecord() {
-
         for(final FlushListener listener: recordEndListener){
             listener.flush(recordCount, currentEntityCount);
         }
@@ -290,14 +288,12 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         dispatch(flattener.getCurrentPath(), "", null);
         currentEntityCount = entityCountStack.pop().intValue();
         flattener.endEntity();
-
     }
 
 
     @Override
     public void literal(final String name, final String value) {
         flattener.literal(name, value);
-
     }
 
     @Override

--- a/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
+++ b/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
@@ -82,8 +82,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
     private static final InterceptorFactory NULL_INTERCEPTOR_FACTORY = new NullInterceptorFactory();
     private static final Map<String, String> NO_VARS = Collections.emptyMap();
 
-    private final Registry<NamedValueReceiver> dataRegistry =
-            new WildcardRegistry<>();
+    private final Registry<NamedValueReceiver> dataRegistry = new WildcardRegistry<>();
     private final List<NamedValueReceiver> elseSources = new ArrayList<>();
 
     private final Map<String, Map<String, String>> maps = new HashMap<>();
@@ -103,7 +102,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
     private boolean elseNestedEntityStarted;
     private String currentLiteralName;
     private static final Logger LOG = LoggerFactory.getLogger(Metamorph.class);
-    final private Pattern literalPatternOfEntityMarker = Pattern.compile(flattener.getEntityMarker(), Pattern.LITERAL);
+
     protected Metamorph() {
         // package private
         init();
@@ -280,6 +279,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         ++entityCount;
         currentEntityCount = entityCount;
         entityCountStack.push(Integer.valueOf(entityCount));
+
         flattener.startEntity(name);
     }
 
@@ -316,11 +316,12 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
     }
 
     private void dispatch(final String path, final String value, final List<NamedValueReceiver> fallbackReceiver, final boolean endEntity) {
-        final List<NamedValueReceiver> matchingData = dataRegistry.get(path);
-        if (matchingData != null && !matchingData.isEmpty()) {
+        final List<NamedValueReceiver> matchingData = getData(path);
+
+        if (matchingData != null) {
             send(path, value, matchingData);
         }
-        else if (fallbackReceiver != null && (literalPatternOfEntityMarker.split(path).length >0 && dataRegistry.get(literalPatternOfEntityMarker.split(path)[0]).isEmpty())) {
+        else if (fallbackReceiver != null) {
             if (endEntity) {
                 if (elseNestedEntityStarted) {
                     outputStreamReceiver.endEntity();
@@ -331,17 +332,25 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
                 final String entityName = elseNested ? flattener.getCurrentEntityName() : null;
 
                 if (entityName != null) {
-                    if (!elseNestedEntityStarted) {
-                        outputStreamReceiver.startEntity(entityName);
-                        elseNestedEntityStarted = true;
+                    if (getData(entityName) == null) {
+                        if (!elseNestedEntityStarted) {
+                            outputStreamReceiver.startEntity(entityName);
+                            elseNestedEntityStarted = true;
+                        }
+
+                        send(currentLiteralName, value, fallbackReceiver);
                     }
-                    send(currentLiteralName, value, fallbackReceiver);
                 }
                 else {
                     send(path, value, fallbackReceiver);
                 }
             }
         }
+    }
+
+    private List<NamedValueReceiver> getData(final String path) {
+        final List<NamedValueReceiver> matchingData = dataRegistry.get(path);
+        return matchingData != null && !matchingData.isEmpty() ? matchingData : null;
     }
 
     private void send(final String path, final String value, final List<NamedValueReceiver> dataList) {

--- a/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
+++ b/metamorph/src/main/java/org/metafacture/metamorph/Metamorph.java
@@ -100,7 +100,8 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
     private int recordCount;
     private final List<FlushListener> recordEndListener = new ArrayList<>();
     private boolean elseNested;
-    final private Pattern literalPatternOfEntityMarker = Pattern.compile(flattener.getEntityMarker(), Pattern.LITERAL);
+    private boolean elseNestedEntityStarted;
+    private String currentLiteralName;
     private static final Logger LOG = LoggerFactory.getLogger(Metamorph.class);
 
     protected Metamorph() {
@@ -202,7 +203,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         flattener.setReceiver(new DefaultStreamReceiver() {
             @Override
             public void literal(final String name, final String value) {
-                dispatch(name, value, getElseSources());
+                dispatch(name, value, getElseSources(), false);
             }
         });
     }
@@ -252,7 +253,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         final String identifierFinal = identifier;
 
         outputStreamReceiver.startRecord(identifierFinal);
-        dispatch(StandardEventNames.ID, identifierFinal, null);
+        dispatch(StandardEventNames.ID, identifierFinal, null, false);
     }
 
     @Override
@@ -285,7 +286,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     @Override
     public void endEntity() {
-        dispatch(flattener.getCurrentPath(), "", null);
+        dispatch(flattener.getCurrentPath(), "", getElseSources(), true);
         currentEntityCount = entityCountStack.pop().intValue();
         flattener.endEntity();
     }
@@ -293,6 +294,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
 
     @Override
     public void literal(final String name, final String value) {
+        currentLiteralName = name;
         flattener.literal(name, value);
     }
 
@@ -314,37 +316,43 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         outputStreamReceiver.closeStream();
     }
 
-    protected void dispatch(final String path, final String value, final List<NamedValueReceiver> fallbackReceiver) {
-        List<NamedValueReceiver> matchingData = dataRegistry.get(path);
-        boolean fallback = false;
-        if (matchingData == null || matchingData.isEmpty()) {
-            fallback = true;
-            matchingData = fallbackReceiver;
+    private void dispatch(final String path, final String value, final List<NamedValueReceiver> fallbackReceiver, final boolean endEntity) {
+        final List<NamedValueReceiver> matchingData = dataRegistry.get(path);
+
+        if (matchingData != null && !matchingData.isEmpty()) {
+            send(path, value, matchingData);
         }
-        if (null != matchingData) {
-            send(path, value, matchingData, fallback);
+        else if (fallbackReceiver != null) {
+            if (endEntity) {
+                if (elseNestedEntityStarted) {
+                    outputStreamReceiver.endEntity();
+                    elseNestedEntityStarted = false;
+                }
+            }
+            else {
+                final String entityName = elseNested ? flattener.getCurrentEntityName() : null;
+
+                if (entityName != null) {
+                    if (!elseNestedEntityStarted) {
+                        outputStreamReceiver.startEntity(entityName);
+                        elseNestedEntityStarted = true;
+                    }
+
+                    send(currentLiteralName, value, fallbackReceiver);
+                }
+                else {
+                    send(path, value, fallbackReceiver);
+                }
+            }
         }
     }
 
-    private void send(final String path, final String value, final List<NamedValueReceiver> dataList,
-            final boolean fallback) {
+    private void send(final String path, final String value, final List<NamedValueReceiver> dataList) {
         for (final NamedValueReceiver data : dataList) {
-            String key = path;
-            if (fallback && elseNested) {
-                if (flattener.getCurrentEntityName() != null) {
-                    outputStreamReceiver.startEntity(flattener.getCurrentEntityName());
-                    key = literalPatternOfEntityMarker.split(path)[1];
-                }
-            }
             try {
-                data.receive(key, value, null, recordCount, currentEntityCount);
+                data.receive(path, value, null, recordCount, currentEntityCount);
             } catch (final RuntimeException e) {
                 errorHandler.error(e);
-            }
-            if (fallback && elseNested) {
-                if (flattener.getCurrentEntityName() != null) {
-                    outputStreamReceiver.endEntity();
-                }
             }
         }
     }
@@ -375,7 +383,7 @@ public final class Metamorph implements StreamPipe<StreamReceiver>, NamedValuePi
         }
 
         if (name.length() != 0 && name.charAt(0) == FEEDBACK_CHAR) {
-            dispatch(name, value, null);
+            dispatch(name, value, null, false);
             return;
         }
 

--- a/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ b/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -421,6 +421,114 @@ public final class TestMetamorphBasics {
         );
     }
 
+    // https://github.com/hagbeck/metafacture-sandbox/tree/master/else-nested-entities
+    @Test
+    public void shouldHandleUseCaseSandboxElseNestedEntities() {
+        assertMorph(receiver,
+                "<rules>" +
+                  "<entity name='85640' flushWith='85640.3' reset='true'>" +
+                    "<data source='85640.u' name='u'>" +
+                      "<contains string='hdl.handle.net/2003/' />" +
+                      "<replace pattern='http://' with='https://' />" +
+                    "</data>" +
+                    "<data source='85640.u' name='u'>" +
+                      "<contains string='doi.org/10.17877/' />" +
+                      "<replace pattern='http://' with='https://' />" +
+                      "<replace pattern='dx.doi.org' with='doi.org' />" +
+                    "</data>" +
+                    "<data source='85640.x' name='x' />" +
+                    "<data source='85640.3' name='3' />" +
+                  "</entity>" +
+                  "<entity name='8564 ' flushWith='8564 .3' reset='true'>" +
+                    "<data source='8564 .u' name='u'>" +
+                      "<contains string='hdl.handle.net/2003/' />" +
+                      "<replace pattern='http://' with='https://' />" +
+                    "</data>" +
+                    "<data source='8564 .u' name='u'>" +
+                      "<contains string='doi.org/10.17877/' />" +
+                      "<replace pattern='http://' with='https://' />" +
+                      "<replace pattern='dx.doi.org' with='doi.org' />" +
+                    "</data>" +
+                    "<data source='8564 .x' name='x' />" +
+                    "<data source='8564 .3' name='3' />" +
+                  "</entity>" +
+                  "<data source='_elseNested' />" +
+                "</rules>",
+                i -> {
+                    i.startRecord("ID1687931");
+                    i.literal("leader", "00564nam a2200024 c 4500");
+                    i.literal("001", "1687931");
+                    i.literal("007", "|| ||||||||||||||||||||");
+                    i.literal("008", "||||||nuuuuuuuu|||||| | |||||||||||und||");
+                    i.startEntity("035  ");
+                    i.literal("a", "(UNION_SEAL)HT019953476");
+                    i.endEntity();
+                    i.startEntity("24500");
+                    i.literal("a", "Design and analysis of an asymmetric mutation operator");
+                    i.literal("c", "Thomas Jansen and Dirk Sudholt");
+                    i.endEntity();
+                    i.startEntity("8564 ");
+                    i.literal("z", "Freie Internetressource");
+                    i.endEntity();
+                    i.startEntity("85640");
+                    i.literal("u", "http://hdl.handle.net/2003/22116");
+                    i.literal("x", "Resolving-System");
+                    i.literal("3", "Volltext");
+                    i.endEntity();
+                    i.startEntity("85640");
+                    i.literal("u", "http://dx.doi.org/10.17877/DE290R-14123");
+                    i.literal("x", "Resolving-System");
+                    i.literal("3", "Volltext");
+                    i.endEntity();
+                    i.startEntity("9801 ");
+                    i.literal("e", "HBZ");
+                    i.endEntity();
+                    i.startEntity("997  ");
+                    i.literal("a", "20190130");
+                    i.endEntity();
+                    i.startEntity("9984 ");
+                    i.literal("z", "Freie Internetressource");
+                    i.endEntity();
+                },
+                o -> {
+                    o.get().startRecord("ID1687931");
+                    o.get().literal("leader", "00564nam a2200024 c 4500");
+                    o.get().literal("001", "1687931");
+                    o.get().literal("007", "|| ||||||||||||||||||||");
+                    o.get().literal("008", "||||||nuuuuuuuu|||||| | |||||||||||und||");
+                    o.get().startEntity("035  ");
+                    o.get().literal("a", "(UNION_SEAL)HT019953476");
+                    o.get().endEntity();
+                    o.get().startEntity("24500");
+                    o.get().literal("a", "Design and analysis of an asymmetric mutation operator");
+                    o.get().literal("c", "Thomas Jansen and Dirk Sudholt");
+                    o.get().endEntity();
+                    o.get().startEntity("8564 ");
+                    o.get().literal("z", "Freie Internetressource");
+                    o.get().endEntity();
+                    o.get().startEntity("85640");
+                    o.get().literal("u", "https://hdl.handle.net/2003/22116");
+                    o.get().literal("x", "Resolving-System");
+                    o.get().literal("3", "Volltext");
+                    o.get().endEntity();
+                    o.get().startEntity("85640");
+                    o.get().literal("u", "https://doi.org/10.17877/DE290R-14123");
+                    o.get().literal("x", "Resolving-System");
+                    o.get().literal("3", "Volltext");
+                    o.get().endEntity();
+                    o.get().startEntity("9801 ");
+                    o.get().literal("e", "HBZ");
+                    o.get().endEntity();
+                    o.get().startEntity("997  ");
+                    o.get().literal("a", "20190130");
+                    o.get().endEntity();
+                    o.get().startEntity("9984 ");
+                    o.get().literal("z", "Freie Internetressource");
+                    o.get().endEntity();
+                }
+        );
+    }
+
 
     @Test
     public void shouldMatchCharacterWithQuestionMarkWildcard() {

--- a/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ b/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -87,24 +87,19 @@ public final class TestMetamorphBasics {
 
     @Test
     public void shouldHandleUnmatchedLiteralsAndEntitiesInElseSource() {
-        testElseData(
-                "<rules>" +
-                "    <data source='_else'/>" +
-                "</rules>"
-        );
+        testElseData("_else");
     }
 
     @Test
     public void shouldHandleUnmatchedLiteralsAndEntitiesInElseFlattenedSource() {
-        testElseData(
-                "<rules>" +
-                "    <data source='_elseFlattened'/>" +
-                "</rules>"
-        );
+        testElseData("_elseFlattened");
     }
 
-    private void testElseData(final String morphDef) {
-        assertMorph(receiver, morphDef,
+    private void testElseData(final String elseKeyword) {
+        assertMorph(receiver,
+                "<rules>" +
+                "  <data source='" + elseKeyword + "'/>" +
+                "</rules>",
                 i -> {
                     i.startRecord("1");
                     i.literal("Shikotan", "Aekap");

--- a/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ b/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -356,6 +356,69 @@ public final class TestMetamorphBasics {
     }
 
     @Test
+    public void shouldNotHandleDataByElseNestedSourceIfDataBelonginToEntityIsRuledByMorph() {
+        assertMorph(receiver,
+                "<rules>" +
+                "  <entity name='USA1'>" +
+                "    <data source='USA1.Sylt' name='Hawaii' />" +
+                "  </entity>" +
+                "  <entity name='USA2' sameEntity='true' reset='true' flushWith='USA2' flushIncomplete='true'>" +
+                "    <data source='USA2.Sylt' name='Hawaii' />" +
+                "    <data source='USA2.Langeoog' name='Langeoog' />" +
+                "  </entity>" +
+                "  <entity name='USA3' sameEntity='true' reset='true' flushWith='USA3' flushIncomplete='true'>" +
+                "    <data source='USA3.Sylt' name='Hawaii' />" +
+                "  </entity>" +
+                "  <data source='_elseNested' />" +
+                "</rules>",
+                i -> {
+                    i.startRecord("1");
+                    i.literal("Shikotan", "Aekap");
+                    i.startEntity("Germany");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.startEntity("USA1");
+                    i.literal("Sylt", "Aloha");
+                    i.endEntity();
+                    i.startEntity("USA2");
+                    i.literal("Sylt", "Aloha");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.startEntity("USA2");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.startEntity("USA3");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.endRecord();
+                },
+                (o, f) -> {
+                    o.get().startRecord("1");
+                    o.get().literal("Shikotan", "Aekap");
+                    o.get().startEntity("Germany");
+                    o.get().literal("Langeoog", "Moin");
+                    o.get().literal("Baltrum", "Moin Moin");
+                    o.get().endEntity();
+                    o.get().startEntity("USA1");
+                    o.get().literal("Hawaii", "Aloha");
+                    o.get().endEntity();
+                    o.get().startEntity("USA2");
+                    o.get().literal("Hawaii", "Aloha");
+                    o.get().literal("Langeoog", "Moin");
+                    o.get().endEntity();
+                    o.get().startEntity("USA2");
+                    o.get().literal("Langeoog", "Moin");
+                    o.get().endEntity();
+                    o.get().endRecord();
+                }
+        );
+    }
+
+
+    @Test
     public void shouldMatchCharacterWithQuestionMarkWildcard() {
         assertMorph(receiver,
                 "<rules>" +

--- a/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ b/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -294,8 +294,12 @@ public final class TestMetamorphBasics {
                     i.endRecord();
                 },
                 (o, f) -> {
+                    // Pass-through coordinates with morph whether to start/end an entity
                     final boolean coordinatesWithEntity = false;
+
+                    // Pass-through and morph entities are separated (one ends when the other starts)
                     final boolean separatesFromEntity = false;
+
                     o.get().startRecord("1");
                     o.get().literal("Shikotan", "Aekap");
                     o.get().startEntity("Germany");
@@ -308,8 +312,8 @@ public final class TestMetamorphBasics {
                     o.get().startEntity("USA2");
                     o.get().literal("Hawaii", "Aloha");
                     if (!coordinatesWithEntity) {
-                    o.get().endEntity();
-                    o.get().startEntity("USA2");
+                        o.get().endEntity();
+                        o.get().startEntity("USA2");
                     }
                     o.get().literal("Langeoog", "Moin");
                     o.get().literal("Baltrum", "Moin Moin");
@@ -317,19 +321,19 @@ public final class TestMetamorphBasics {
                     o.get().startEntity("USA3");
                     o.get().literal("Langeoog", "Moin");
                     if (!coordinatesWithEntity) {
-                    o.get().startEntity("USA3");
+                        o.get().startEntity("USA3");
                     }
                     else if (separatesFromEntity) {
-                    o.get().endEntity();
-                    o.get().startEntity("USA3");
+                        o.get().endEntity();
+                        o.get().startEntity("USA3");
                     }
                     o.get().literal("Hawaii", "Aloha");
                     if (!coordinatesWithEntity) {
-                    o.get().endEntity();
+                        o.get().endEntity();
                     }
                     else if (separatesFromEntity) {
-                    o.get().endEntity();
-                    o.get().startEntity("USA3");
+                        o.get().endEntity();
+                        o.get().startEntity("USA3");
                     }
                     o.get().literal("Baltrum", "Moin Moin");
                     o.get().endEntity();
@@ -337,18 +341,18 @@ public final class TestMetamorphBasics {
                     o.get().literal("Langeoog", "Moin");
                     o.get().literal("Baltrum", "Moin Moin");
                     if (!coordinatesWithEntity) {
-                    o.get().startEntity("USA4");
+                        o.get().startEntity("USA4");
                     }
                     else if (separatesFromEntity) {
-                    o.get().endEntity();
-                    o.get().startEntity("USA4");
+                        o.get().endEntity();
+                        o.get().startEntity("USA4");
                     }
                     o.get().literal("Hawaii", "Aloha");
                     if (!coordinatesWithEntity) {
-                    f.apply(2).endEntity();
+                        f.apply(2).endEntity();
                     }
                     else {
-                    o.get().endEntity();
+                        o.get().endEntity();
                     }
                     o.get().endRecord();
                 }
@@ -356,17 +360,17 @@ public final class TestMetamorphBasics {
     }
 
     @Test
-    public void shouldNotHandleDataByElseNestedSourceIfDataBelonginToEntityIsRuledByMorph() {
+    public void shouldNotHandleDataByElseNestedSourceIfDataBelongingToEntityIsRuledByMorph() {
         assertMorph(receiver,
                 "<rules>" +
                 "  <entity name='USA1'>" +
                 "    <data source='USA1.Sylt' name='Hawaii' />" +
                 "  </entity>" +
-                "  <entity name='USA2' sameEntity='true' reset='true' flushWith='USA2' flushIncomplete='true'>" +
+                "  <entity name='USA2' sameEntity='true' flushWith='USA2'>" +
                 "    <data source='USA2.Sylt' name='Hawaii' />" +
                 "    <data source='USA2.Langeoog' name='Langeoog' />" +
                 "  </entity>" +
-                "  <entity name='USA3' sameEntity='true' reset='true' flushWith='USA3' flushIncomplete='true'>" +
+                "  <entity name='USA3' sameEntity='true' flushWith='USA3'>" +
                 "    <data source='USA3.Sylt' name='Hawaii' />" +
                 "  </entity>" +
                 "  <data source='_elseNested' />" +
@@ -395,7 +399,7 @@ public final class TestMetamorphBasics {
                     i.endEntity();
                     i.endRecord();
                 },
-                (o, f) -> {
+                o -> {
                     o.get().startRecord("1");
                     o.get().literal("Shikotan", "Aekap");
                     o.get().startEntity("Germany");

--- a/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
+++ b/metamorph/src/test/java/org/metafacture/metamorph/TestMetamorphBasics.java
@@ -98,6 +98,10 @@ public final class TestMetamorphBasics {
     private void testElseData(final String elseKeyword) {
         assertMorph(receiver,
                 "<rules>" +
+                "  <entity name='Germany'>" +
+                "    <data source='Germany.Sylt' name='Hawaii' />" +
+                "    <data source='Germany.Borkum' />" +
+                "  </entity>" +
                 "  <data source='" + elseKeyword + "'/>" +
                 "</rules>",
                 i -> {
@@ -105,6 +109,8 @@ public final class TestMetamorphBasics {
                     i.literal("Shikotan", "Aekap");
                     i.startEntity("Germany");
                     i.literal("Langeoog", "Moin");
+                    i.literal("Sylt", "Aloha");
+                    i.literal("Borkum", "Tach");
                     i.endEntity();
                     i.startEntity("Germany");
                     i.literal("Baltrum", "Moin Moin");
@@ -115,7 +121,92 @@ public final class TestMetamorphBasics {
                     o.get().startRecord("1");
                     o.get().literal("Shikotan", "Aekap");
                     o.get().literal("Germany.Langeoog", "Moin");
+                    o.get().startEntity("Germany");
+                    o.get().literal("Hawaii", "Aloha");
+                    o.get().literal("Germany.Borkum", "Tach");
+                    o.get().endEntity();
                     o.get().literal("Germany.Baltrum", "Moin Moin");
+                    o.get().endRecord();
+                }
+        );
+    }
+
+    @Test
+    public void issue338_shouldPreserveSameEntitiesInElseNestedSource() {
+        assertMorph(receiver,
+                "<rules>" +
+                "  <data source='_elseNested' />" +
+                "</rules>",
+                i -> {
+                    i.startRecord("1");
+                    i.literal("lit1", "val1");
+                    i.startEntity("ent1");
+                    i.literal("lit2", "val2");
+                    i.literal("lit3", "val3");
+                    i.endEntity();
+                    i.literal("lit4", "val4");
+                    i.startEntity("ent2");
+                    i.literal("lit5", "val5");
+                    i.literal("lit6", "val6");
+                    i.literal("lit7", "val7");
+                    i.endEntity();
+                    i.startEntity("ent2"); // sic!
+                    i.literal("lit8", "val8");
+                    i.literal("lit9", "val9");
+                    i.endEntity();
+                    i.endRecord();
+                    i.startRecord("2");
+                    i.startEntity("ent1");
+                    i.literal("lit1", "val1");
+                    i.literal("lit2", "val2");
+                    i.endEntity();
+                    i.startEntity("ent2");
+                    i.literal("lit3", "val3");
+                    i.literal("lit4", "val4");
+                    i.literal("lit5", "val5");
+                    i.literal("lit6", "val6");
+                    i.endEntity();
+                    i.startEntity("ent3");
+                    i.literal("lit7", "val7");
+                    i.literal("lit8", "val8");
+                    i.endEntity();
+                    i.literal("lit9", "val9");
+                    i.endRecord();
+                },
+                o -> {
+                    o.get().startRecord("1");
+                    o.get().literal("lit1", "val1");
+                    o.get().startEntity("ent1");
+                    o.get().literal("lit2", "val2");
+                    o.get().literal("lit3", "val3");
+                    o.get().endEntity();
+                    o.get().literal("lit4", "val4");
+                    o.get().startEntity("ent2");
+                    o.get().literal("lit5", "val5");
+                    o.get().literal("lit6", "val6");
+                    o.get().literal("lit7", "val7");
+                    o.get().endEntity();
+                    o.get().startEntity("ent2");
+                    o.get().literal("lit8", "val8");
+                    o.get().literal("lit9", "val9");
+                    o.get().endEntity();
+                    o.get().endRecord();
+                    o.get().startRecord("2");
+                    o.get().startEntity("ent1");
+                    o.get().literal("lit1", "val1");
+                    o.get().literal("lit2", "val2");
+                    o.get().endEntity();
+                    o.get().startEntity("ent2");
+                    o.get().literal("lit3", "val3");
+                    o.get().literal("lit4", "val4");
+                    o.get().literal("lit5", "val5");
+                    o.get().literal("lit6", "val6");
+                    o.get().endEntity();
+                    o.get().startEntity("ent3");
+                    o.get().literal("lit7", "val7");
+                    o.get().literal("lit8", "val8");
+                    o.get().endEntity();
+                    o.get().literal("lit9", "val9");
                     o.get().endRecord();
                 }
         );
@@ -147,13 +238,118 @@ public final class TestMetamorphBasics {
                     o.get().literal("Shikotan", "Aekap");
                     o.get().startEntity("Germany");
                     o.get().literal("Langeoog", "Moin");
-                    o.get().endEntity();
-                    o.get().startEntity("Germany");
                     o.get().literal("Baltrum", "Moin Moin");
                     o.get().endEntity();
                     o.get().startEntity("USA");
                     o.get().literal("Hawaii", "Aloha");
                     o.get().endEntity();
+                    o.get().endRecord();
+                }
+        );
+    }
+
+    @Test
+    public void shouldHandlePartiallyUnmatchedLiteralsAndEntitiesInElseNestedSource() {
+        assertMorph(receiver,
+                "<rules>" +
+                "  <entity name='USA1'>" +
+                "    <data source='USA1.Sylt' name='Hawaii' />" +
+                "  </entity>" +
+                "  <entity name='USA2'>" +
+                "    <data source='USA2.Sylt' name='Hawaii' />" +
+                "  </entity>" +
+                "  <entity name='USA3'>" +
+                "    <data source='USA3.Sylt' name='Hawaii' />" +
+                "  </entity>" +
+                "  <entity name='USA4'>" +
+                "    <data source='USA4.Sylt' name='Hawaii' />" +
+                "  </entity>" +
+                "  <data source='_elseNested' />" +
+                "</rules>",
+                i -> {
+                    i.startRecord("1");
+                    i.literal("Shikotan", "Aekap");
+                    i.startEntity("Germany");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.startEntity("USA1");
+                    i.literal("Sylt", "Aloha");
+                    i.endEntity();
+                    i.startEntity("USA2");
+                    i.literal("Sylt", "Aloha");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.startEntity("USA3");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Sylt", "Aloha");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.endEntity();
+                    i.startEntity("USA4");
+                    i.literal("Langeoog", "Moin");
+                    i.literal("Baltrum", "Moin Moin");
+                    i.literal("Sylt", "Aloha");
+                    i.endEntity();
+                    i.endRecord();
+                },
+                (o, f) -> {
+                    final boolean coordinatesWithEntity = false;
+                    final boolean separatesFromEntity = false;
+                    o.get().startRecord("1");
+                    o.get().literal("Shikotan", "Aekap");
+                    o.get().startEntity("Germany");
+                    o.get().literal("Langeoog", "Moin");
+                    o.get().literal("Baltrum", "Moin Moin");
+                    o.get().endEntity();
+                    o.get().startEntity("USA1");
+                    o.get().literal("Hawaii", "Aloha");
+                    o.get().endEntity();
+                    o.get().startEntity("USA2");
+                    o.get().literal("Hawaii", "Aloha");
+                    if (!coordinatesWithEntity) {
+                    o.get().endEntity();
+                    o.get().startEntity("USA2");
+                    }
+                    o.get().literal("Langeoog", "Moin");
+                    o.get().literal("Baltrum", "Moin Moin");
+                    o.get().endEntity();
+                    o.get().startEntity("USA3");
+                    o.get().literal("Langeoog", "Moin");
+                    if (!coordinatesWithEntity) {
+                    o.get().startEntity("USA3");
+                    }
+                    else if (separatesFromEntity) {
+                    o.get().endEntity();
+                    o.get().startEntity("USA3");
+                    }
+                    o.get().literal("Hawaii", "Aloha");
+                    if (!coordinatesWithEntity) {
+                    o.get().endEntity();
+                    }
+                    else if (separatesFromEntity) {
+                    o.get().endEntity();
+                    o.get().startEntity("USA3");
+                    }
+                    o.get().literal("Baltrum", "Moin Moin");
+                    o.get().endEntity();
+                    o.get().startEntity("USA4");
+                    o.get().literal("Langeoog", "Moin");
+                    o.get().literal("Baltrum", "Moin Moin");
+                    if (!coordinatesWithEntity) {
+                    o.get().startEntity("USA4");
+                    }
+                    else if (separatesFromEntity) {
+                    o.get().endEntity();
+                    o.get().startEntity("USA4");
+                    }
+                    o.get().literal("Hawaii", "Aloha");
+                    if (!coordinatesWithEntity) {
+                    f.apply(2).endEntity();
+                    }
+                    else {
+                    o.get().endEntity();
+                    }
                     o.get().endRecord();
                 }
         );

